### ISSUE TITLE
Problem with the "purchase" event

### ISSUE THREAD

**Issue by atoupet-toki** [duracelltomi/gtm4wp#320]
Hello!
We have a problem with the "purchase" event.
Wordpress and all plugins are updated. We have also tested gtm4wp version 1.20-beta.2.
We also followed the tutorials to setup the plugins and the GTM (New empty container, and importing the json container from the tutorial).

We have all events in dataLayer, except "purchase".
We've debugged the plugin code, and it seems that the `$order->update_meta_data` functions are called before the purchase_datalayer. If we comment out the 2 lines `$order->update_meta_data( '_ga_tracked', 1 );` it works. But we can see that .... the 2 lines are after the purchase_datalayer.

We are using Complianz plugin for cookies. But same probleme without this plugin
Has anyone submitted a similar request? Do you need more information to test and validate something? Please let me know.

Thanks in advance

Here is capture of datalayer:
![Capture d’écran, le 2024-01-30 à 11 34 39](https://github.com/duracelltomi/gtm4wp/assets/38693082/f10320ee-038f-40cf-958a-382f37102373)

**Comment by duracelltomi** [comment]
Hi,

In similar cases there was some sort of payment processing plugin that abnormally executed codes normally run once during generating the order received page but in the background. Therefore when the order received page was really loaded, the order was already set to be tracked once.

**Comment by takibulislam** [comment]
Hi, 
im dealing with the same issue,  using default woocommerce checkout and payment processing options, but I have no clue now what exactly happening there, 

is there any possible solution?

**Comment by duracelltomi** [comment]
@takibulislam do you have a testing/sandbox/dev version of your site where I can place an order without the need to actually pay for the products?

**Comment by takibulislam** [comment]
Hi @duracelltomi
im not a web developer for this website, but I can get a 100% discount code with free shipping so you can avoid paying anything, do you think It will work?

**Comment by takibulislam** [comment]
@duracelltomi  
i think the info i have provided was incorrect, here is the payment configuration of our website. 
![image](https://github.com/duracelltomi/gtm4wp/assets/54490333/21410bc6-e8b6-4a01-851e-7f951044e84a)

this is the replay I'v got from the Dev team,
* it is using Afterpay Gateway for WooCommerce and Square
* and PayPal Gateway By Payment Plugins

please let me know if you need anything else,  appreciate your help

**Comment by duracelltomi** [comment]
Yes, it could work but it could happen that I will need you to patch some GTM4WP files to test some theories. That will require a separate environment.

**Comment by duracelltomi** [comment]
Is it working if you select the direct bank transfer during checkout?

**Comment by takibulislam** [comment]
![image](https://github.com/duracelltomi/gtm4wp/assets/54490333/08fd7ec8-fcf4-42fd-be1a-ce8b1e7c3aea)
I have just placed an order with Direct Bank transfer, it's not working, unfortunately. 

I can provide you a duplicate version of our website, if you want please let me know how can I send you the info

**Comment by takibulislam** [comment]
@duracelltomi 
Hello, 
we have created a test version of our website, if you want to test I can send you the information. please let me know,

Thank you.

**Comment by duracelltomi** [comment]
Please send details to messenger at gtm4wp dot com. Thanks!

**Comment by takibulislam** [comment]
@duracelltomi 
I have sent login info to the email you mentioned. I would love to know if there is any update on it. 

seems like many people facing the same issue nowadays, hope it will be resolved for all of us. 

Thanks for your support and waiting for your response!

**Comment by duracelltomi** [comment]
I received your message. I have a tough day today, it can happen that I will only be able to deep dive tomorrow

**Comment by duracelltomi** [comment]
OK, I was just too curious :-)
When I turn on the following GTM4WP option: 'do not flag orders as being tracked' - the purchase event is there.

It means that for any reason, another plugin is calling codes normally executing while showing the user the order confirmation page. GTM4WP flags the order as being tracked once and when the real order received page is shown, no purhcase event is fired since GTM4WP "sees" that it was already tracked once.

**Comment by atoupet-toki** [comment]
This is exactly the same problem I mentioned, but it doesn't seem to me to be related to a payment processing plugin, we don't have the same @takibulislam and me.

**Comment by takibulislam** [comment]
@atoupet-toki In my case I also feel like payment processing isn't responsible, I have tried turning off that plugin but still faced the same issue. but I'm not 100% sure.

but anyway I hope Thomas will be able to figure out a decent solution for us.

**Comment by duracelltomi** [comment]
I just want to let you know that I am committed finding a solution to this situation but I have limited time to investigate. I do my best to find a solution as soon as possible.

**Comment by duracelltomi** [comment]
Hi,

No yet but I had an extremely busy week behind me, most probably I will have time to continue with the investigation next week.

**Comment by Blueboots** [comment]
Adding my 2c here as well as in the other similar thread.

In my case, as well as no purchase event on the order-received page, the 'ecommerce' object is not generated, and so no ecommerce data is pulled into the data layer variables. There is an 'orderData' object which I guess is what Woocommerce makes.

I wondered if @takibulislam and @atoupet-toki had the same issue.
